firebase-sessions/src/main/kotlin/com/google/firebase/sessions/SessionLifecycleClient.kt

```diff
@@ -0,0 +1,181 @@
+          super.handleMessage(msg)
+        }
+      }
+    }
+
+    fun handleSessionUpdate(sessionId: String) {
+      Log.i(TAG, "Session update received: $sessionId")
+      curSessionId = sessionId
+    }
+  }
+
+  /** The connection object to the [SessionLifecycleService]. */
+  private val serviceConnection =
+    object : ServiceConnection {
+      override fun onServiceConnected(className: ComponentName, serviceBinder: IBinder) {
+        Log.i(TAG, "Connected to SessionLifecycleService. Queue size ${queuedMessages.size}")
+        service = Messenger(serviceBinder)
+        serviceBound = true
+        val messagesToSend = ArrayList<Message>()
+        queuedMessages.drainTo(messagesToSend)
+        sendLifecycleEvents(messagesToSend)
+      }
+
+      override fun onServiceDisconnected(className: ComponentName) {
+        Log.i(TAG, "Disconnected from SessionLifecycleService")
+        service = null
+        serviceBound = false
+      }
+    }
+
+  /**
+   * Binds to the [SessionLifecycleService] and passes a callback [Messenger] that will be used to
+   * relay session updates to this client.
+   */
+  fun bindToService(appContext: Context): Unit {
+    Intent(appContext, SessionLifecycleService::class.java).also { intent ->
+      Log.i(TAG, "Binding service to application.")
+      // This is necessary for the onBind() to be called by each process
+      intent.setAction(android.os.Process.myPid().toString())
+      intent.putExtra(
+        SessionLifecycleService.CLIENT_CALLBACK_MESSENGER,
+        Messenger(ClientUpdateHandler())
+      )
+      appContext.bindService(
+        intent,
+        serviceConnection,
+        Context.BIND_IMPORTANT or Context.BIND_AUTO_CREATE
+      )
+    }
+  }
+
+  /**
+   * Should be called when any activity in this application process goes to the foreground. This
+   * will relay the event to the [SessionLifecycleService] where it can make the determination of
+   * whether or not this foregrounding event should result in a new session being generated.
+   */
+  fun foregrounded(activity: Activity): Unit {
+    sendLifecycleEvent(SessionLifecycleService.FOREGROUNDED)
+  }
+
+  /**
+   * Should be called when any activity in this application process goes from the foreground to the
+   * background. This will relay the event to the [SessionLifecycleService] where it will be used to
+   * determine when a new session should be generated.
+   */
+  fun backgrounded(activity: Activity): Unit {
+    sendLifecycleEvent(SessionLifecycleService.BACKGROUNDED)
+  }
+
+  /**
+   * Sends a message to the [SessionLifecycleService] with the given event code. This will
+   * potentially also send any messages that have been queued up but not successfully delivered to
+   * thes service since the previous send.
+   */
+  private fun sendLifecycleEvent(messageCode: Int) {
+    val allMessages = ArrayList<Message>()
+    queuedMessages.drainTo(allMessages)
```

Comment:
Maybe this line and the line above could be extracted to a method `drainQueuedToNewList()`? You do it a couple of times.

Reply:
Done